Add the state of obtained permission scopes to the Google account hook

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Part of #1000

:warning: After the discussions of https://github.com/woocommerce/google-listings-and-ads/pull/1060#issuecomment-951954104 and https://github.com/woocommerce/google-listings-and-ads/pull/1060#issuecomment-953442883, this PR only added the state of obtained scopes to the Google account hook. The feature of preview existing GMC account was reverted.

~To preview the existing Google Merchant Center accounts when user's Google is connected but didn't grant required permissions (aka *scopes* in Google term), and reduce the chance of development conflicts with the UX improvements project.~ This PR:

- add the state of obtained scopes to the Google account hook
- ~add `disabled` prop to the `<ConnectMCCard>` for the preview mode~
- ~adjust some logic to preview existing Google Merchant Center accounts if it could~

### Screenshots:

:warning: This video is outdated.

https://user-images.githubusercontent.com/17420811/138687363-518063f3-2795-483b-9ef4-89ecb1413b15.mp4

### Detailed test instructions:

:warning: This test instructions are outdated.

1. Disconnect all accounts.
2. Go to the onboarding setup and proceed with the account connection setup.
3. When authorizing plugin with your Google account, only tick off the "**Manage your product listings and accounts for Google Shopping**" permission.
4. It should display your existing Google Merchant Center accounts with disabled states on the card UI of Googe Merchant Center.
5. Redo steps 1~2. If accept the "**Manage your product listings and accounts for Google Shopping**" and "**Manage your new site verifications with Google**" permissions, it should be able to continue the next steps as before.

### Changelog entry

eason9487 committed:
Add the state of obtained scopes to the Google account  hook

eason9487 committed:
Add hasFinishedResolution state to the useExistingGoogleMCAccounts hook

eason9487 committed:
Add disabled prop to the ConnectMCCard for the preview mode

eason9487 committed:
Preview existing GMC accounts if Google is connected but has no required scopes

eason9487 committed:
Avoid overlapping opacity multiple times on disabled elements of select and button.

ecgan commented on js/src/hooks/useGoogleAccount.js:
```diff
@@ -9,6 +9,29 @@ import { useSelect } from '@wordpress/data';
+	CONTENT: 'https://www.googleapis.com/auth/content',
+	// Manage new site verifications with Google
+	SITE_VERIFICATION:
+		'https://www.googleapis.com/auth/siteverification.verify_only',
```
❔  I think we should make the key name more specific, for example `SITE_VERIFICATION_VERIFY_ONLY`. If we just use `SITE_VERIFICATION`, we might get it confused with the `https://www.googleapis.com/auth/siteverification` scope. See more info below:

- https://developers.google.com/identity/protocols/oauth2/scopes#siteverification
- https://developers.google.com/site-verification/v1/getting_started#OAuth2Authorizing

eason9487 replied:
Renamed by https://github.com/woocommerce/google-listings-and-ads/pull/1060/commits/8b43c5417e96760f73205dec5b2d1ceff3d11ddf

ecgan commented on js/src/hooks/useGoogleAccount.js:
```diff
@@ -9,6 +9,29 @@ import { useSelect } from '@wordpress/data';
+		scopes.includes( SCOPE.CONTENT ) &&
+		scopes.includes( SCOPE.SITE_VERIFICATION );
+
+	state.allRequired = state.gmcRequired && state.adsRequired;
```
❔ Thinking we probably don't need `allRequired`? 🤔  And also thinking do we really need the `Required` suffix? 😄

eason9487 replied:
> Thinking we probably don't need `allRequired`?

The `allRequired` will be used on the Google account reconnection page. If a user had set up the Google Ads account, it would ask the user to grant all required permissions when reconnecting.

> do we really need the `Required` suffix?

Consider they are boolean type. With `Required` suffix, it should be easier to infer the meaning of true/false directly from the naming when seeing them in other places.

ecgan commented on js/src/hooks/useGoogleAccount.js:
```diff
@@ -9,6 +9,29 @@ import { useSelect } from '@wordpress/data';
 import { STORE_KEY } from '.~/data/constants';
 import useJetpackAccount from './useJetpackAccount';
 
+const SCOPE = {
```
❔ Consider using [`Object.freeze`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze) on this scope object?

Example: https://github.com/woocommerce/google-listings-and-ads/blob/6770e91b66f5af563cac3d9b101cea5a18fca06f/js/src/data/utils.js#L19-L26

eason9487 replied:
The `SCOPE` doesn't been exported. I think it should be ok without wrapping `Object.freeze`.

ecgan commented on js/src/setup-mc/setup-stepper/setup-accounts/index.js:
```diff
@@ -36,7 +36,12 @@ const SetupAccounts = ( props ) => {
+		isGoogleConnected && scope.gmcRequired
+	);
+	const maybePreviewExistingGoogleMCAccounts =
+		isGoogleConnected && ! scope.gmcRequired;
```
💅  I'm not a fan of the "maybe" prefix, it makes it more difficult to understand and reason about. Looking into all the "maybe" usage here, I'm thinking we can have a slightly more verbose naming to make it clearer and more explicit. For example, we can rename this `isGoogleConnectedInsufficientGMCScope`, or just drop this variable because `isGoogleConnected && ! scope.gmcRequired` is quite clear for me.

eason9487 replied:
Renamed by https://github.com/woocommerce/google-listings-and-ads/pull/1060/commits/ed4424e1190652409f12f9a734adfb7245a9710b

ecgan commented on js/src/setup-mc/setup-stepper/setup-accounts/index.js:
```diff
@@ -63,7 +68,12 @@ const SetupAccounts = ( props ) => {
+				disabled={ isGoogleMCAccountDisabled }
+				maybePreviewExistingAccounts={
+					maybePreviewExistingGoogleMCAccounts
+				}
```
💅  This prop `maybePreviewExistingAccounts` can be renamed to something like `disabledShowAccounts` to convey the meaning of "load and show the accounts drop down list even when disabled is true". 

Incorporating my other comment in this file, this can become...

```suggestion
				disabledShowAccounts={
					isGoogleConnected && ! scope.gmcRequired
				}
```

... which IMO is easier to reason about. 

I name it `disabledShowAccounts` so that it would show up next to `disabled` props in editor's intellisense suggestion in alphabetical order.

eason9487 replied:
I was stuck in this naming for a while when developing. 😵 

> This prop `maybePreviewExistingAccounts` can be renamed to something like `disabledShowAccounts` to convey the meaning of "load and show the accounts drop down list even when disabled is true".

Consider that we are trying to display a disabled UI with existing connectable Google Merchant Center accounts in a special state, which fulfills only if a Google account is connected without granting required permissions. I thought "maybe" is a proper prefix to indicate that this prop is very special.

However, the `disabledShowAccounts` naming looks more like "do not show accounts" to me, and it doesn't even point out that "accounts" is stood for connected or existing accounts. After comparing them, I prefer keeping it as is.

ecgan commented on js/src/setup-mc/setup-stepper/setup-accounts/google-mc-account/section-content.js:
```diff
@@ -2,24 +2,47 @@
+			return <MaybePreviewConnectMCCard />;
+		}
 		return <DisabledCard />;
 	}
```
💅  

- If `isNonConnected` is true, we display `MaybePreviewConnectMCCard` which is a disabled card with accounts drop down list.
- If `isNonConnected` is false (which means MC account is connected), we display `DisabledCard` which is a disabled card with a text `'Connect your Merchant Center'`.

I'm thinking we can actually simplify these by just showing `DisabledCard`. We don't really need `MaybePreviewConnectMCCard`, it is not really useful to display disabled card with accounts drop down list.

eason9487 replied:
> I'm thinking we can actually simplify these by just showing `DisabledCard`. We don't really need `MaybePreviewConnectMCCard`, it is not really useful to display disabled card with accounts drop down list.

Looks like it should be a 📜  item. 🤔 

It's implemented based on the project requirements. The follow-up: https://github.com/woocommerce/google-listings-and-ads/pull/1060#issuecomment-951954104

eason9487 replied:
The preview feature was reverted by https://github.com/woocommerce/google-listings-and-ads/pull/1060/commits/499a2e2a4460ffcc2034e227f22ced27e771fe6b. See https://github.com/woocommerce/google-listings-and-ads/pull/1060#issuecomment-953442883.

ecgan commented on js/src/setup-mc/setup-stepper/setup-accounts/google-mc-account/section-content.js:
```diff
@@ -2,24 +2,47 @@
+import ConnectMCCard from './connect-mc-card';
 import SpinnerCard from '.~/components/spinner-card';
 
+const MaybePreviewConnectMCCard = () => {
```
💅  A better non-"maybe" naming like `DisabledShowAccountsCard`. 😄

eason9487 replied:
The preview feature was reverted by https://github.com/woocommerce/google-listings-and-ads/pull/1060/commits/499a2e2a4460ffcc2034e227f22ced27e771fe6b. See https://github.com/woocommerce/google-listings-and-ads/pull/1060#issuecomment-953442883.

ecgan reviewed (approved):
Tested as per test instructions, works as expected. I left quite a few comments on naming (especially on `SITE_VERIFICATION` and the use of "maybe" prefix in naming), and I'm going to post a separate comment on the UI design, which hopefully can simplify the UI code more. Approving for now since there isn't anything major blocking us.

ecgan commented:
@j111q , referring to the screenshot video posted by @eason9487 in this PR description, after users connect their Google account with insufficient scope granted, we will display a disabled Google MC section card with accounts drop down list in it:

![image](https://user-images.githubusercontent.com/417342/138889113-ddf678b1-5700-44b7-bc34-e6c51b94fcc8.png)

IMO, it isn't really that useful to display the disabled card with accounts drop down, because the users can't interact with them anyway. It also works a bit slower because it is making API call, and you can see the loading state at 0:18 second in the video.

![image](https://user-images.githubusercontent.com/417342/138890867-4b809de5-1155-41c0-be5f-24421d862e28.png)

Suggestion: What do you think if we just show the following disabled "Connect your merchant center" card, when users have not connected their Google accounts with all the Merchant Center required permissions? From the technical code perspective, that should also make things simpler.

![image](https://user-images.githubusercontent.com/417342/138890089-25f0827d-f81b-4c48-8617-1a678f244cd9.png)

(Note to @eason9487 : this should help to address the "maybe" naming prefix in my review comments. 😄 )

eason9487 committed:
Rename scope constant of Google account

Address https://github.com/woocommerce/google-listings-and-ads/pull/1060#discussion_r736365386

eason9487 committed:
Rename a variable to isGoogleConnectedInsufficientGMCScope

Address https://github.com/woocommerce/google-listings-and-ads/pull/1060#discussion_r736488541

j111q commented:
> IMO, it isn't really that useful to display the disabled card with accounts drop down, because the users can't interact with them anyway. It also works a bit slower because it is making API call, and you can see the loading state at 0:18 second in the video.

That's fair enough @ecgan, good catch! Let's go with your suggestion. Thank you!! 🙏

eason9487 committed:
Revert "Rename a variable to isGoogleConnectedInsufficientGMCScope"

This reverts commit ed4424e1190652409f12f9a734adfb7245a9710b.

eason9487 committed:
Revert a range commits about previewing existing GMC accounts.

This reverts commit range c85e11c68a54e66fb3abe26c8544894c55d015c5^..741eb6f8109b677f391685793d0076a47617bf38.

eason9487 commented:
After reverting, the remaining changes have been reviewed and approved before. Will proceed to merge.
